### PR TITLE
Expand docs for eps

### LINKED ISSUE #20542: Documentation for eps(x::AbstractFloat)
The documentation for eps(x::AbstractFloat) is currently worded as

> The distance between `x` and the next larger representable floating-point value of the same
`DataType` as `x`.

which sounds to me like it's defined as  `_eps = x-> nextfloat(x) - x`, which it's not.

```julia
julia> eps(prevfloat(Inf))
1.99584030953472e292

julia> eps(-Inf)
NaN
```
Though for every other value I could think of, `eps` and `_eps` seem to behave the same.

### PR DESCRIPTION
**Pull request by simonbyrne**
Fixed #20542 and #13447.

### CODE REVIEW

**Comment by KristofferC** [comment]
Of course not required, but more illustrative doctests is always nice for us "learn by example" people :)

**Comment by simonbyrne** [comment]
Good idea: any suggestions?

**Comment by KristofferC** [comment]
Usually with doctest I firstly write the most obvious things and then try to show on some "interesting" cases.

One thing that could be showed is maybe:

```julia
julia> 1.0 + eps(1.0)
1.0000000000000002

julia> 1.0 + eps(1.0)/2
1.0
```

File: base/float.jl [hunk 1]
```diff
@@ -711,6 +711,35 @@ end
...
+default). This is defined as the gap between 1 and the next largest value representable by
+`T`, and is equivalent to `eps(one(T))`.
+"""
+eps{T<:AbstractFloat}(::Type{T})
+
+"""
+    eps(x::AbstractFloat)
+
+Returns the *unit in last place* (ulp) of `x`. This is the distance between consecutive
+representable floating point values at `x` of the same type as `x`. In most cases, if `x`
```
Comment (tkelman) [inline base/float.jl]: this `x` at the end of the line doesn't look right

File: base/float.jl [hunk 2]
```diff
@@ -733,9 +747,28 @@ the distance on either side is different, then the larger of the two is taken, t
 with the exception of the smallest and largest finite values (`nextfloat(-Inf)` and
 `prevfloat(Inf)`), which round to the smaller of the values.
 
-The rationale for this choice is that under default `RoundNearest` behaviour, if ``y`` is
-a real number which rounds to `x`, then the absolute difference between ``y`` and `x` is
+The rationale for this choice is that under default `RoundNearest` behaviour, for any
+realy number ``y`` which rounds to `x`, the absolute difference between ``y`` and `x` is
```
Comment (KristofferC) [inline base/float.jl]: `realy`

File: base/float.jl [hunk 3]
```diff
@@ -733,9 +747,28 @@ the distance on either side is different, then the larger of the two is taken, t
...
+4.440892098500626e-16
+
+julia> x = prevfloat(Inf)  # largest finite Float64
+1.7976931348623157e308
+
+julia> x + eps(x)/2            # rounds up
+Inf
+
+julia> x + prevfloat(eps(x)/2) # rounds down
+1.7976931348623157e308
```
Comment (KristofferC) [inline base/float.jl]: missing backticks

**Comment by KristofferC** [comment]
Very nice, thanks for adding that

File: base/float.jl [hunk 4]
```diff
@@ -711,6 +711,69 @@ end
     eps() = eps(Float64)
 end
 
+"""
+    eps{T<:AbstractFloat}(::Type{T})
```
Comment (ararslan) [inline base/float.jl]: Should this be `eps(::Type{T}) where T<:AbstractFloat`?

File: base/float.jl [hunk 5]
```diff
@@ -711,6 +711,69 @@ end
...
+Returns the *unit in last place* (ulp) of `x`. This is the distance between consecutive
+representable floating point values at `x` of the same type as `x`. In most cases, if the
+distance on either side of `x` is different then the larger of the two is taken, that is
+
+    eps(x) == max(x-prevfloat(x), nextfloat(x)-x)
+
+with the exception of the smallest and largest finite values (`nextfloat(-Inf)` and
+`prevfloat(Inf)`), which round to the smaller of the values.
+
+The rationale for this choice is that under default `RoundNearest` behaviour, for any
```
Comment (ararslan) [inline base/float.jl]: If I'm not mistaken, the docs mostly use American English, so I think "behaviour" should be "behavior."

Comment (tkelman) [inline base/float.jl]:
```
$ git grep behaviour | wc -l
20

$ git grep behavior | wc -l
171
```

File: base/float.jl (hunk 4 above)
Comment (nalimilan) [inline base/float.jl]: Even better, `eps(::Type{<:AbstractFloat})`.

Comment (simonbyrne) [inline base/float.jl]: Probably best to keep the `T`, as I refer to it in the text.

File: base/float.jl [hunk 6]
```diff
@@ -739,17 +739,22 @@ eps{T<:AbstractFloat}(::Type{T})
...
+The exceptions to this rule are the smallest and largest finite values
+(e.g. `nextfloat(-Inf)` and `prevfloat(Inf)` for `Float64`), which round to the smaller of
+the values.
 
-The rationale for this choice is that under default `RoundNearest` behaviour, for any
-real number ``y`` which rounds to `x`, the absolute difference between ``y`` and `x` is
-bounded by half of `eps(x)`.
+The rationale for this behavior is that `eps` bounds the floating point rounding
+error. Under the default `RoundNearest` rounding mode, if ``y`` is a real number and ``x``
+is the nearest floating point number ``y``, then
```
Comment (ararslan) [inline base/float.jl]: nearest floating point number _to_ y?

Comment (simonbyrne) [inline base/float.jl]: i think you're looking at an old version.

Comment (ararslan) [inline base/float.jl]: Ah, so I am. Sorry!

File: base/float.jl [hunk 7]
```diff
@@ -711,6 +711,74 @@ end
...
+1.1920929f-7
+
+julia> 1.0 + eps()
+1.0000000000000002
+
+julia> 1.0 + eps()/2
+1.0
+```
+"""
+eps{T<:AbstractFloat}(::Type{T})
```
Comment (ararslan) [inline base/float.jl]: I think this should also be `eps(::Type{T}) where T<:AbstractFloat`, or `eps(::Type{<:AbstractFloat})`

File: doc/src/stdlib/base.md [hunk 8]
```diff
@@ -97,8 +97,7 @@ Base.realmin
 Base.realmax
 Base.maxintfloat
 Base.sizeof(::Type)
-Base.eps(::Union{Type{BigFloat},Type{Float64},Type{Float32},Type{Float16}})
-Base.eps()
+Base.eps{T<:AbstractFloat}(::Type{T})
```
Comment (ararslan) [inline doc/src/stdlib/base.md]: Same here, `where` or `<:AbstractFloat`

File: base/float.jl [hunk 9]
```diff
@@ -711,6 +711,74 @@ end
     eps() = eps(Float64)
 end
 
+"""
+    eps(::Type{T}) where T<:AbstractFloat
```
Comment (StefanKarpinski) [inline base/float.jl]: `eps(::Type{<:AbstractFloat})`?

Comment (tkelman) [inline base/float.jl]: T is mentioned in the text so should be a named parameter

File: base/float.jl [hunk 10]
```diff
@@ -711,6 +711,74 @@ end
     eps() = eps(Float64)
 end
 
+"""
+    eps(::Type{<:AbstractFloat})
```
Comment (ararslan) [inline base/float.jl]: The `where` was better actually, since as Tony mentioned the parameter `T` is referenced in the explanation.

**Pull request merged**